protokube/pkg/protokube/ali_volume.go

```diff
@@ -0,0 +1,322 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package protokube
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/denverdino/aliyungo/common"
+	"github.com/denverdino/aliyungo/ecs"
+	"github.com/denverdino/aliyungo/metadata"
+	"github.com/golang/glog"
+	"k8s.io/kops/protokube/pkg/etcd"
+	"k8s.io/kops/protokube/pkg/gossip"
+	gossipali "k8s.io/kops/protokube/pkg/gossip/ali"
+	"k8s.io/kops/upup/pkg/fi/cloudup/aliup"
+)
+
+// ALIVolumes is the Volumes implementation for Aliyun ECS
+type ALIVolumes struct {
+	client *ecs.Client
+
+	clusterTag string
+	region     string
+	zone       string
+	instanceId string
+	internalIP net.IP
+}
+
+var _ Volumes = &ALIVolumes{}
+
+func NewALIVolumes() (*ALIVolumes, error) {
+	accessKeyId := os.Getenv("ALIYUN_ACCESS_KEY_ID")
+	if accessKeyId == "" {
+		return nil, fmt.Errorf("error initialing ALIVolumes: ALIYUN_ACCESS_KEY_ID cannot be empty")
+	}
+	accessKeySecret := os.Getenv("ALIYUN_ACCESS_KEY_SECRET")
+	if accessKeySecret == "" {
+		return nil, fmt.Errorf("error initialing ALIVolumes: ALIYUN_ACCESS_KEY_SECRET cannot be empty")
+	}
+	ecsEndpoint := os.Getenv("ALIYUN_ECS_ENDPOINT")
+	if ecsEndpoint == "" {
+		// TODO: shall we raise error here?
```

Comment:
If you have a default, I'd say probably not, but you might log e.g. `glog.Infof("Defaulting ALIYUN_ECS_ENDPOINT to %s", ecsEndpoint)`